### PR TITLE
[WIP] [LLM] Speculative decoding: refine batching strategy

### PR DESCRIPTION
**Pull request by xiangyuT**
## Description

Use attention_mask and padding to refine batching strategy.

### CODE REVIEW

**Comment by qiyuangong** [comment]
PR is not clean. Don't include other author's commit. 
Please resolve conflict with current main.

File: python/llm/src/bigdl/llm/transformers/speculative.py [hunk 1]
```diff
@@ -115,6 +115,27 @@ def clear_benchmarks(self):
     self.n_drafted = 0
     self.n_matched = 0
 
+    self.draft_nums = []
+    self.accept_nums = []
+    self.n_drafteds = []
+    self.n_matcheds = []
+    self.actual_iters = [0] * 100
```
Comment (qiyuangong) [inline python/llm/src/bigdl/llm/transformers/speculative.py:122]: magic number 100

File: python/llm/src/bigdl/llm/transformers/speculative.py [hunk 2]
```diff
@@ -115,6 +115,27 @@ def clear_benchmarks(self):
...
+
+
+def _update_attention_mask(attention_mask, len=1):
+    return torch.cat(
+        [attention_mask, attention_mask.new_ones((attention_mask.shape[0], len))],
+        dim=-1)
+
+def _update_attention_mask_left(attention_mask, len=1):
```
Comment (qiyuangong) [inline python/llm/src/bigdl/llm/transformers/speculative.py:130]: Maybe we can merge 3 functions into 1

File: python/llm/src/bigdl/llm/transformers/speculative.py [hunk 3]
```diff
@@ -243,18 +264,24 @@ def _update_past_key_values_storage_cpu(self, past_key_values, past_key_values_s
...
+
 @torch.no_grad()
 def speculative_generate(self,
                          inputs: Optional[torch.Tensor] = None,
                          draft_model=None,
                          max_new_tokens=10,
                          max_step_draft=8,
                          th_stop_draft=0.8,
-                         auto_th_stop_draft=True,
+                         auto_th_stop_draft=False,
```
Comment (qiyuangong) [inline python/llm/src/bigdl/llm/transformers/speculative.py:279]: Should be True by default.

File: python/llm/src/bigdl/llm/transformers/speculative.py [hunk 4]
```diff
@@ -297,7 +324,25 @@ def speculative_generate(self,
...
+    model_kwargs["output_hidden_states"] = generation_config.output_hidden_states
+    # decoder-only models with inputs_embeds forwarding must use caching
+    # (otherwise we can't detect whether we are generating the first new token or not,
+    # and we only want to use the embeddings for the first new token)
+    if not self.config.is_encoder_decoder and model_input_name == "inputs_embeds":
+        model_kwargs["use_cache"] = True
+    else:
+        model_kwargs["use_cache"] = generation_config.use_cache
+
+    accepts_attention_mask = "attention_mask" in set(
```
Comment (qiyuangong) [inline python/llm/src/bigdl/llm/transformers/speculative.py:337]: accepted_attention_mask

File: python/llm/src/bigdl/llm/transformers/speculative.py [hunk 5]
```diff
@@ -297,7 +324,25 @@ def speculative_generate(self,
...
+        inspect.signature(self.forward).parameters.keys())
+    requires_attention_mask = "encoder_outputs" not in model_kwargs
```
Comment (qiyuangong) [inline python/llm/src/bigdl/llm/transformers/speculative.py:339]: required_attention_mask

File: python/llm/src/bigdl/llm/transformers/speculative.py [hunk 6]
```diff
@@ -368,6 +414,18 @@ def speculative_generate(self,
     e2e_tic = 0.0
 
     self.clear_benchmarks()
+    attention_mask = model_kwargs["attention_mask"]
+
+    if len(input_ids.shape) == 1:
+        batch_size = 1
+    else:
+        batch_size = input_ids.size(0)
+    pad_token_id = generation_config.pad_token_id
```
Comment (qiyuangong) [inline python/llm/src/bigdl/llm/transformers/speculative.py:423]: When bs==1, we don't need padding and mask. I think we should separated bs==1 and bs!=1.

File: python/llm/src/bigdl/llm/transformers/speculative.py [hunk 7]
```diff
@@ -381,27 +439,42 @@ def speculative_generate(self,
...
             output = self(input_ids=current_input_ids,
                           past_key_values=past_key_values,
                           attention_mask=attention_mask,
+                          position_ids=position_ids,
                           return_dict=True,
                           use_cache=True)
             logits = output['logits']
             logits = logits[:, -1:]
-            logits[:, -1, :] = logits_processor(current_input_ids, logits[:, -1, :])
+            # logits[:, -1, :] = logits_processor(current_input_ids, logits[:, -1, :])
```
Comment (qiyuangong) [inline python/llm/src/bigdl/llm/transformers/speculative.py:468]: Do we need to handle logits here?
